### DISCUSSION TITLE
Copy node (visible)

### DISCUSSION THREAD

**Discussion by Carsten-Kaftan** [freeplane/freeplane#1495] [Announcements]
Filtering nodes is one of the big strengths of Freeplane – to further use the results, I miss the feature »Copy node (visible)«, which would only copy the ancestors (for instance after selecting the root), descendants and nodes, which are shown and not filtered.
By the way, »Copy node (visible)« would also be useful to copy the current nodes of interest into another map, without dragging along every other node of the selected branches.

**Comment by dpolivaev** [comment]
I converted this issue to discussion because I want to ask community to choose between different implementation options:

- I can add a new action to the menu and call Copy visible nodes. I am not sure if I like the suggested name though.
- I can modify the existing action so that it does not copy unfiltered nodes.
- I can add the user option "Copies nodes hidden by filter" defining how the existing copy action behaves.

Given the shortcut control+C is bound to the existing action, which way do you prefer and why?

**Comment by Carsten-Kaftan** [comment]
Ignoring nodes hidden by filter when copying seems to me the most »natural« behaviour, so I prefer the second proposition (»modify the existing action so that it does not copy unfiltered nodes«). The amount of unlooked-for content in the clipboard, after copying a filtered map, is always an unwelcome surprise; I see no use case for including rejected nodes into the copy. To get more content, you could activate show ancestors/descendants or disable the filter.

**Comment by sifran-github** [comment]
I think that the first option (keeping the default behavior and adding another "copy-only-visible-nodes" as a menu option) is the only logical option and doing otherwise is dangerous. having "copy-only-visible-nodes" by default Ctrl+C and Ctrl+X could lead to devastating results, and I will try to explain why:

Lets imagine a very common scenario: The user has a lot of nodes, and he just want to search(filter) for some node and copy/cut it to another map. If this "copy-only-visible-nodes" by default is implemented, he has to be very carful before he is doing it... after finding the node that he wants to copy/cut by filtering the map, he must make sure that: (1) the filter include descendants, and that (2) he uncheck view->hidden nodes... (and maybe more reasons that a node might be hidden?)

he has to be very careful before doing this simple copy/cut and paste operation.
If he forgets to do any of these safety actions before the operation .... HE WILL ACTUALLY LOSE DATA.

It is much better to avoid losing data by default and allowing the user to "copy-only-visible-nodes" as an additional menu option.

**Reply by sifran-github** [reply]
and if someone doesn't agree, then please answer: how CTRL+X should behave?

1. When pressing CTRL+X on a node that has some hidden descendants, the user will automatically lose all the hidden nodes forever?
2. or CTRL+X will copy to the clipboard different content than by using CTRL+C ?

These 2 options do not sound reasonable to me.

**Reply by sifran-github** [reply]
I like the idea of "Copy node (only visible)" but as an additional menu option.

**Reply by sifran-github** [reply]
@dpolivaev please read this commend carefully.

If this change is implemented, it will actually become dangerous to copy-paste or cut-paste in Freeplane, and can lead to loss of data.

Even when doing copy-paste, since you know that you copied some nodes to another location, you mentally know that you can now delete these nodes from the original location.
But if change is implemented, you would be mistaken to believe so, because it would result in more content being deleted from the original location, and the copied nodes in the new location will not contain all the content... and until you'll find about it, it will be too late and your data will be lost forever!

**Reply by sifran-github** [reply]
I really don't understand why introducing such a risky behavior when a simple solution as "Copy node (only visible)" menu option can be introduced, and if people want to use it often, they can set their own shortcut for it.

I even don't see any reason to put a default keyboard shortcut for it. I personally will rarely use this option, but it can be "nice to have".

But again, please do not implement it as a default behavior. Too risky.

**Comment by Carsten-Kaftan** [comment]
There certainly is a point in this, especially when changing default behaviour. But:

- That copying works in another way than cutting is not this unusual in many contexts, for example when handling read-only-data, or generating new identifiers when copying but keeping identifiers when cutting.
- I would not cut in a map with filtered nodes, too much fear to create havoc, so my answer to the question »how CTRL+X should behave?« is a very clear »Cutting should not work in a filtered map.« When moving nodes I often get the message »All nodes must have the same parent to use this function.« (although it would be possible to move them independently); perhaps something similar would be acceptable for the courageous users who try to cut what they don't see.

Apart from this, for me personally the more options the merrier; there always arrives the situation when the extra option saves a lot of time – so an additional menu option may be less elegant but gives the users more precise control about what exactly should be done.

**Reply by dpolivaev** [reply]
I think that "cut" should always ignore the filter because I do not see why another behaviour might be better.

**Reply by sifran-github** [reply]
> I think that "cut" should always ignore the filter because I do not see why another behaviour might be better.

I agree.
And Ctrl+X should copy to the clipboard whatever Ctrl+C is copying. Otherwise it is weird.
And as I explained above, it can be dangerous and can lead to loss of data.

**Comment by dpolivaev** [comment]
I have discovered further inconsistencies in the current logic and I am not sure which fix is better.
When content is copied to the clipboard, Freeplane prepares different representations namely plain text, HTML, and the internal format used when the paste action is called by Freeplane itself.

At the moment, filtered nodes are not copied to text or html, but they are copied into the internal format.

Actually, after the fix, it should be consistent again. Also, achieving consistency would require changing the current logic.

There is also another case: when nodes are copied using drag and drop, internally it also uses the copying logic. And if I introduce a second action, I have to decide, which of the two should be attached to drag and drop.

Another solution would be to show the user a dialog with both options and a "remember the choice" checkbox when copying is triggered by the action or by drag-and-drop. Actually, I think that this is the safest way to resolve the conflicts because the users would be shown the dialog at least once and so they have a chance of making a conscious choice.

**Reply by Carsten-Kaftan** [reply]
Wouldn't drag-and-drop rather be cut-and-paste than copy-and-paste? When cut always ignores the filter, as you suggested, implementing drag-and-drop with cut instead of copy perhaps could be a straightforward solution.

**Reply by dpolivaev** [reply]
If you hold extra key, I do not know if it is control or alt or option, I am so confused by MacBooks, then it performs a copy

**Reply by Carsten-Kaftan** [reply]
Ah yes, I did not think of this. Obviously it is much easier to nag about a missing feature (me) than to do the real work, and maintain an application (you).

**Reply by sifran-github** [reply]
in my opinion:

- CTRL+C, CTRL+X = always copies everything to the clipboard. Also the external representations, such as text or HTML, should be full and not ignore filtered nodes.
- "Copy node (only visible)" menu option = this is the option to use in order to copy visible nodes only. To the internal format and to the external representations as well.
- Drag and drop: Always copies everything. I don't really understand what does it even mean to drag and drop without the hidden/unfiltered nodes. So when I drag some node to another location, will it automatically delete all the hidden/filtered subnodes and I will lose them forever? just by this simple drag and drop operation? Why would anybody want such risky drag and drop operation? even drag-and-drop to copy (instead of move) that will drop some nodes, is crazy risky as I explained already.

Am I missing anything?
![image](https://github.com/freeplane/freeplane/assets/5730355/78e8f836-a4b2-4fbf-bbed-3ffe7c822d78)

**Reply by dpolivaev** [reply]
Because the current behavior (copying all nodes for internal paste and only filtered nodes for export) seems to be inconsistent but it is useful,
Because this feature request is the first time ever a user wants to copy only filtered nodes,
I think that the best solution is introducing some new options in the preferences instead of a new action. The default values are indicated by x.

````text
Apply Filter When Copying Nodes for Paste as Nodes [ ]
Apply Filter When Copying Nodes for Paste as Text [x]
````

**Reply by dpolivaev** [reply]
@Carsten-Kaftan would it satisfy your needs?
@sifran-github Is this in line with your concerns?

**Reply by sifran-github** [reply]
@dpolivaev What's wrong with my suggestion? I gave a solution which is:

- Safe (most important! we don't want users to lose data)
- Logical
- Consistent (CTRL+X always copies the same data as CTRL+C, and also internal representation always contain the same nodes as the external representation (text, html))
- Allows anything that the user wants to be done ("a user wants to copy only filtered nodes" - no problem, use the menu option to copy only filtered nodes).

I actually do not see any reason to implement it in any other way.

**Reply by sifran-github** [reply]
and if some users really want to suffer, they can change the default CTRL+C to execute the new menu option (only visible).

This is risky and will lead to them losing their data in the future.... but.... it was their responsibly to choose such a risky option. They can also bind the CTRL+C to a "Delete" function if they choose to... :)

**Reply by dpolivaev** [reply]
> What's wrong with my suggestion?

I would not call it "wrong" because I believe humans should be very careful about using this strong and absolute word. Actually, one needs to be omniscient to use it correctly.

I prefer the solution I described because it allows us to keep the current logic which I believe makes sense in most cases as a default so that on the one side, no one gets surprised by a sudden change, and on the other side, copying nodes as text would still apply the filter, and it is what I believe satisfies the most real use cases. On the same side, it adds flexibility by configuration if users have other needs.

**Reply by sifran-github** [reply]
@dpolivaev I'm sorry but I humbly disagree. I actually checked now and you are right, when I copy nodes that have filtered descendants, and I paste them in two different locations, I get different results:
1. Pasting inside a mind map, pasts everything.
2. Pasting inside a text editor, pasts only the filtered nodes.

This inconsistency is a bug, not a feature.
It is super weird behavior from a software.
If I would discover it by myself, I would post an issue report on it.

This can also lead to data loss.
For example, if someone copies some data, checks that he copied everything by pasting it into a temporary mind map, and then actually pasting it into some text file in order to back up his data.... he will not be aware that he is actually not pasting all the data.

it is even worse if pasting into mindmap would behave like that, but this inconsistency is bad also.

With my suggestion (which was the suggestion in the original post as well), it will be very clear what the user is doing.

**Reply by dpolivaev** [reply]
I am not sure. Maybe you are right in some cases, maybe not. 
Until now nobody complained, and therefore I believe that the behavior was satisfying almost everyone.
Other users seem not to have an interest in the topic.
Anyway, I have offered everyone an opportunity to bring their arguments, and all arguments have been heard.
Therefore I have implemented it in a way I think is the most flexible.
I do not think that software should always protect users from themselves. At least in a non-default configuration users should be allowed to choose what they need.

**Reply by sifran-github** [reply]
> At least in a non-default configuration users should be allowed to choose what they need.

I agree with this statement.
I mostly worry about the default implementation and the cases where it can lead to data-loss.

Since data-loss probably worst thing that can happen to a user, I try to defend the position of safety by default.

**Reply by dpolivaev** [reply]
For me keeping the existing defaults seems to be the right decision. However, it would be easy to change until it goes stable, if there was a broad agreement in the user community for the change.

**Reply by sifran-github** [reply]
@dpolivaev  I'm very happy that you didn't add the most dangerous part, which was to make it default to copy filtered nodes when pasting into freeplane.

But still, in my opinion. the solution is not good enough and the inconsistency between copied data for mind map and copied data for text, is still a bug in freeplane...

With the current solution, I will still need to remember that the default configuration is dangerous if I want to copy content outside of freeplane, and I have to make sure that I copied everything.

It is so inconsistent that I still have to test what Ctrl+X does and what Ctrl+C does because at the moment they do different things... I'm sorry but this is a very strange behavior.

and I still didn't hear any argument against my solution. @Carsten-Kaftan can do exactly what he wants with the simple implementation of an additional "copy node (only visible)" menu. He can even add a keyboard shortcut to it, if he wants to.

@Carsten-Kaftan , will this way of implementing it (adding a menu item, that you can customize its keyboard shortcut) be good for your needs?

**Comment by Carsten-Kaftan** [comment]
> @Carsten-Kaftan would it satisfy your needs? @sifran-github Is this in line with your concerns?

Yes, I'm very fine with this solution. 
It is a little surprising, that I am the first user ever who not just wants to view the results of filtering, but also to use the results for further processing. Having the results of meticulous filtering locked in unrecoverable in the current map is a sore spot for a long time (»just look, don't touch« is frustrating, here I prefer »what you see is what you get« :-).
Having been a database administrator for 18 years I detest datalosses, but in this special case I feel confident that I am not the only one who can cope with the risks of not copying the whole structure by default. Offering the choice between expert modus (»Apply filter«) and the safe kindergarden modus (as default) would solve the problem, many thanks for offering this.

**Comment by dpolivaev** [comment]
The solution is available in preview 1.8.11_06, 
https://sourceforge.net/projects/freeplane/files/freeplane%20preview/.
You can find the new options at
Preferences…->Behaviour->Apply filter when copying nodes as map nodes
Preferences…->Behaviour->Apply filter when copying nodes as text

and in the command search dialog

**Reply by Carsten-Kaftan** [reply]
After some testing I can confirm that filtering/copying works in version freeplane_1.11.8-pre06~upstream-1_all.deb after activating the option.
That will save me many hours; the map I process has nearly 45000 nodes, from which to extract those of interest for the given situations.
That was unbelievable quick, one week from enhancement-request to pre-version!
🙇 Dear Lightspeed-Dimitry, a very impressing stunt, tons of thanks!!
